hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpc/Socket.java

```diff
@@ -0,0 +1,214 @@
+package com.hazelcast.internal.tpc;
+
+import com.hazelcast.internal.tpc.logging.TpcLogger;
+import com.hazelcast.internal.tpc.logging.TpcLoggerLocator;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.hazelcast.internal.tpc.util.Preconditions.checkNotNull;
+
+/**
+ * The socket for TPC engine.
+ */
+public abstract class Socket implements Closeable {
```

Comment:
I think we should avoid user implemented classes having same name as Java classes (java.net.Socket).
```suggestion
public abstract class CloseableSocket implements Closeable {
```